### DISCUSSION TITLE
Multistate models with mrgsolve

### DISCUSSION THREAD

**Discussion by TDmetric** [metrumresearchgroup/mrgsolve#1040] [Q&A]
A recently published paper in ASCPT PSP (https://ascpt.onlinelibrary.wiley.com/doi/abs/10.1002/cpt.2838) used mrgsolve to simulate from a multistate model with 5 states. I often use mrgsolve to simulate from time-to-event models (a multistate model with 2 states), but am not sure how I could incorporate additional states. Does anyone have any thoughts or code they could share to enlighten me on this? This has been on my mind all week. Thank you in advance!

TD

**Comment by Mutaz94** [comment]
Have you tried using `[ des ]` block to implement differential equations? You should be able to simulate the system by writing down the 5 equations they specified. Solving these equations calculates the probability for each state. 

```

[ des ]

dxdt_s1 = something;
dxdt_s2 = somethingElse; 

[ error ] 
double p_s1 = s1; //probability of state1
double p_s2 = s2; //probability of state2
double p_cu1 = p_s1; //cumulative  prob
double p_cu2 = p_s1 + p_s2; 
```

Does that make sense?

**Reply by TDmetric** [reply]
Mutaz, thanks so much for the response! Yes, this makes sense. The ODEs for these models are surprisingly straightforward. My difficulty is in simulating events from these models. For time to event (TTE) models, a common practice is to draw a random number from a uniform distribution U(0,1) for each patient. When the survival probability drops below the random number, an event is observed (simulated) for the patient. No problem when there are only two states.

Say we have an illness-death model (S1, S2, S3), where S1 is the starting state, S2 is an intermediate state, and S3 is a terminal state. If you're simulating with mrgsolve and a transition from S1 to S2 is simulated, how do you get the hazard (and thus the survival probability) for transitioning from S2 to S3 to start only after the transition from S1 to S2 has occurred? I'm sure this is in part the reason the authors in the paper above calculate the cumulative probability for combinations of states (i.e., CUP1 = A(1), CUP2 = A(1) + A(2), CUP3 = A(1) + A(2) + A(3)), but I can't quite connect the dots yet. 

Thanks in advance for your input! I'm really into these multistate models right now. They're great for joint models and competing risks.

TD

**Comment by smouksassi** [comment]
some useful info might be found here:
https://www.r-bloggers.com/2018/12/simulating-multi-state-models-with-r/

you might want to follow the logic and internals of mstate::mssample()

**Comment by kylebaron** [comment]
Hi @TDmetric  - 

In `mrgsolve`, you can draw the random number like this

```r
[ plugin ] Rcpp autodec

[ error ] 
x = R::runif(0,1); 
```

I think you can use the CUPs to get the state:

```r
[ error ]  
x = R::runif(0,1);

if(x <= CUP1) STATE = 1;
if(x > CUP1 && x <= CUP2) STATE=1;
```

We've done some multistate modeling with nonmem / mrgsolve. I'm hoping to get a shareable example out some time that would show this. 

Kyle

**Reply by TDmetric** [reply]
Thanks so much for your response @kylebaron! I've got several of these multistate models running now. A shareable example would be great though! I'd be really interested to see an example of a joint model with PK/PD and a multistate system. Resetting the multistate compartments with each observation without resetting the system for the PK/PD is tricky. Any advice?
